pkg/controller/daemon/daemon_controller.go

```diff
@@ -1236,6 +1236,15 @@ func (dsc *DaemonSetsController) syncDaemonSet(ctx context.Context, key string)
 		return dsc.updateDaemonSetStatus(ctx, ds, nodeList, hash, false)
 	}
 
+	defer func() {
```

Comment:
This is masking the original error, isn't it?

Reply:
What happens with the error in line1250 if `dsc.manage` fails?

Reply:
I would avoid a complex defer function if possible, on the long term makes the code harder to maintain, since code after the defer has to take that into account and is easier to forget

Reply:
I think we should be okay on the masking part here https://go.dev/play/p/iilbYnmJwbR.

I suppose that is true, but it seems to me that this fits the defer nicely here. Since we would like to update the status no matter what happens in the rest of the function.

I guess we could write a comment that we wish to set the `syncErr` only in case there is none, or somehow make the if conditions more readable?

Reply:
I have similar concerns as @aojea here, and using `defer` is confusing. What if we change the flow of this function such that we wrap these 3 calls here: https://github.com/kubernetes/kubernetes/blob/cac53883f4714452f3084a22e4be20d042a9df33/pkg/controller/daemon/daemon_controller.go#L1239-L1259 in a single function and then invoke `dsc.updateDaemonSetStatus(ctx, ds, nodeList, hash, true)` afterwards always and based on error from the wrapped functions and error from update status we can easily return. The code will be much more readable and easier to understand. Even if any part will grow with features/bug fixes/etc. it will still maintain its original flow. 

PS. I have similar concerns about the fix in https://github.com/kubernetes/kubernetes/pull/109694 where I'd be inclined to change the flow in a similar pattern.

Reply:
> I suppose that is true, but it seems to me that this fits the defer nicely here. Since we would like to update the status no matter what happens in the rest of the function.

agree. I think this is the best option we have.

wrote some comments.

Reply:
I was not sure what the scope of the PR should be, but now I agree changing the flow will be a better and long term soluti0on

Reply:
@soltysh 
updated.

I would make another PR for the statefulset. Is it ok?